operators/api/v1alpha2/labtemplate_types.go

```diff
@@ -0,0 +1,91 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type LabType string
+type EnvironmentType string
+
+const (
+	TypeGUI        LabType         = "GUI"
+	TypeCLI        LabType         = "CLI"
+	ClassContainer EnvironmentType = "Container"
+	ClassVM        EnvironmentType = "VirtualMachine"
+)
+
+// TemplateSpec defines the desired state of Template
+type TemplateSpec struct {
+	WorkspaceRef    GenericRef    `json:"workspace.crownlabs.polito.it/WorkspaceRef,omitempty"`
+	PrettyName      string        `json:"prettyName,omitempty"`
+	Description     string        `json:"description,omitempty"`
+	EnvironmentList []Environment `json:"environmentList,omitempty"`
+	// +kubebuilder:validation:Pattern="^[0-9]+[mhd]$"
+	DeleteAfter string `json:"deleteAfter,omitempty"`
+}
+
+// TemplateStatus defines the observed state of Template
+type TemplateStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make lab-template" to regenerate code after modifying this file
+}
+
+type Environment struct {
+	Name string `json:"name,omitempty"`
+	// +kubebuilder:validation:Enum="GUI";"CLI"
+	LabType   `json:"labType,omitempty"`
+	Resources EnvironmentResources `json:"resources,omitempty"`
+	// +kubebuilder:validation:Enum="VirtualMachine";"Container"
+	EnvironmentType EnvironmentType `json:"environmentType"`
+	Persistent      bool            `json:"persistent"`
+	Image           string          `json:"image"`
+}
+
+type EnvironmentResources struct {
+	CPUNumber   uint32            `json:"cpuNumber,omitempty"`
```

Comment:
Do we want to consider only integer numbers or allow also things like 500m?
Probably kubevirt requires it to be integer, but we could round up in that case

Second, I would add an upper bound to avoid excessive values

Third, probably these fields should not have the omitempty annotation, since they are mandatory

Reply:
I left "omitempty" because it is not required or required for containers. Totally agree for having a upper bound.

Reply:
Ok, I understand.
I would make it nonetheless mandatory in this API to force the requests to be specified, but it is just an opinion